File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<pair<int,int>> stk; //Height,index
        int area = 0; 

        for(int i=0;i<=heights.size();++i){
            int h=(i==heights.size())?0:heights[i];
            int start=i;
            while(!stk.empty() && h<stk.top().first){
                int height=stk.top().first, j=stk.top().second;
                int width=i-j;
                area=max(area,height*width);
                stk.pop();
                start=j;
            }
            stk.push({h,start});
        }
        return area;
    }
};

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        stk = [] #idx height
        area = 0

        for i in range(n+1):
            start = i
            h = heights[i] if i!=n else 0 # To collect final area from the heights remaining in the stack
            while stk and h<stk[-1][1]:
                height = stk[-1][1]
                width = i-stk[-1][0]
                area = max(area,height*width)
                start = stk[-1][0]
                stk.pop()
            stk.append((start,h))
        
        return area